### ISSUE TITLE
`when Foo` wrongly accepts non-bool Foo

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#8681]
when accepts pretty much anything:

```nim
when defined(case2):
  proc fun()=
    type Foo=object
    when Foo: echo "z1"
    elif "XDG_CONFIG_HOME":
      echo "z2"
    else:
      echo "z3"
  fun()
```

I diagnosed this as root cause of https://github.com/nim-lang/Nim/issues/8491 (XDG_CONFIG_HOME was being happily ignored in `getConfigDir`) (which I'm fixing in https://github.com/nim-lang/Nim/pull/8680)

**Comment by tim-st** [comment]
Yes, it accepts too much, [here](https://github.com/nim-lang/Nim/issues/8603#issuecomment-412461315) are other examples with non-bool.
